# Тема 3. Основи Python: синтаксис, типи даних, введення і виведення

## 1. Типова структура програми Python

Програма на мові Python – це текстовий файл з розширенням `.py`, що містить набір інструкцій, які інтерпретатор виконує послідовно, рядок за рядком, згори донизу. Хоча прості програми можуть складатися лише з кількох рядків коду, для кращої організації та читабельності варто дотримуватися певної структури.

### 1.1 Загальні принципи організації коду

**Ключові елементи структури:**

* **Коментарі:** Важлива частина будь-якої програми. Це текстові пояснення, які ігноруються інтерпретатором, але слугують для пояснення логіки коду програмістам. У Python однорядкові коментарі починаються із символу `#`.

In [1]:
# Це коментар - текст після # ігнорується інтерпретатором
# Коментарі використовуються для пояснення коду
# x = 1
# print(x)

In [2]:
"""
Це багаторядковий коментар
Він може займати декілька рядків
"""
'''
y = 2
print(x + y)
'''
print('Привіт, Світ!')

Привіт, Світ!


**Інструкції (Statements):** Кожен рядок коду, що виконує певну дію (наприклад, присвоєння значення змінній, виклик функції), є інструкцією.

Типова структура простої програми може виглядати так:
1.  **Імпорт модулів (за потреби):** Підключення зовнішніх бібліотек, які розширюють функціональність.
2.  **Оголошення констант та глобальних змінних (за потреби):** Визначення даних, що будуть доступні в усій програмі.
3.  **Визначення функцій (за потреби):** Створення блоків коду для повторного використання.
4.  **Основний блок виконання:** Головна логіка програми, що виконується при запуску файлу.

Для найпростіших скриптів програма може складатись лише з основного блоку виконання.
Python не потребує обов’язкової функції `main`, як у деяких інших мовах. Програма починає виконання з першої інструкції у файлі.
Отже, програма Python може виконуватися як скрипт (без функції `main`) або з використанням умовної конструкції `if __name__ == "__main__":` для розмежування основного коду від імпортованого.

In [3]:
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Документація модуля (docstring)
Опис призначення програми
"""

# Імпорт необхідних модулів
import math

# Константи (пишуться великими літерами)
PI = math.pi
RADIUS = 10

# Визначення функцій
def calculate_area(radius):
    """Обчислює площу кола"""
    return PI * radius ** 2

# Основний код програми
def main():
    """Головна функція програми"""
    area = calculate_area(RADIUS)
    print(f"Площа кола: {area:.2f}")

# Точка входу в програму
if __name__ == "__main__":
    main()

Площа кола: 314.16


Рядки
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
```
не впливають на логіку програми, але забезпечують правильне виконання та інтерпретацію файлу в різних середовищах.
1. **Shebang-рядок**: `#!/usr/bin/env python3`
Призначення: Цей рядок називається "shebang" (від "sharp" # і "bang" !) і використовується в Unix-подібних системах (Linux, macOS) для вказівки, яким інтерпретатором виконувати файл.
2. **Кодування**: `# -*- coding: utf-8 -*-`
Призначення: Цей рядок вказує Python, в якому кодуванні записаний файл з кодом.

**Історичний контекст**:
- У Python 2 за замовчуванням використовувалось ASCII-кодування
- Якщо в коді були символи поза ASCII (кирилиця, емодзі, спецсимволи), виникали помилки
- Цей рядок дозволяв явно вказати UTF-8 кодування

**Сучасний стан**:

- У Python 3 UTF-8 є кодуванням за замовчуванням
- Цей рядок більше не є обов'язковим для більшості випадків
 -Залишається корисним для явного вказання кодування або сумісності з Python 2

### 1.2 Синтаксичні особливості Python

**Блоки коду та відступи (Indentation):** На відміну від багатьох інших мов, де для визначення блоків коду використовуються фігурні дужки `{}`, у Python для цього слугують **відступи** (пробіли або табуляція). Це одна з ключових синтаксичних особливостей мови. Всі інструкції в межах одного блоку (наприклад, тіло циклу або функції) повинні мати однаковий рівень відступу. Стандартною практикою є використання 4 пробілів для одного рівня вкладеності. Ця особливість робить код більш читабельним і примушує програмістів дотримуватися єдиного стилю оформлення.

In [4]:
# Приклад використання відступів
number = 15
if number > 10:
    print("Число більше 10")  # Відступ 4 пробіли
    print('Це другий рядок блоку')  # Той самий відступ
print("Цей блок не входить до блоку умови")

Число більше 10
Це другий рядок блоку
Цей блок не входить до блоку умови


| Особливість | Опис | Приклад |
|-------------|------|---------|
| **Відступи** | Визначають блоки коду (4 пробіли) | `    print("Відступ")` |
| **Двокрапка** | Завершує заголовки блоків | `if x > 0:` |
| **Регістрочутливість** | `Name` ≠ `name` | `variable` ≠ `Variable` |
| **Без крапки з комою** | В кінці рядка не потрібна | `x = 10` |

Інструкції в Python зазвичай розміщуються по одній на рядок. Для продовження довгого рядка на наступний використовується зворотна коса риска `\` або круглі дужки.
Неявне продовження рядків відбувається автоматично всередині круглих `()`, квадратних `[]` або фігурних `{}` дужок. Python розуміє, що вираз ще не завершено, поки не буде закрита відповідна дужка, тому дозволяє переносити код на наступні рядки без додаткових символів. Цей спосіб є більш читабельним та рекомендованим порівняно з використанням зворотної косої риски.

Тепер щодо прикладів, як це працює на практиці:
1. Використання круглих дужок для довгих виразів:

In [5]:
first_variable = 1
second_variable = 2
third_variable = 3
fourth_variable = 4
fifth_variable = 5

# Довгий математичний вираз
result = (first_variable + second_variable +
          third_variable + fourth_variable +
          fifth_variable)

# Довгий виклик функції
print("Це дуже довгий рядок",
      "який ми хочемо розбити",
      "на кілька рядків для кращої читабельності")

Це дуже довгий рядок який ми хочемо розбити на кілька рядків для кращої читабельності


2. Використання квадратних дужок для списків:

In [6]:
# Довгий словник
person = {
    "name": "John Doe",
    "age": 30,
    "city": "Kyiv",
    "occupation": "Developer"
}

4. Порівняння з використанням зворотної косої риски:

In [7]:
# Менш читабельний спосіб
result = first_variable + second_variable + \
         third_variable + fourth_variable + \
         fifth_variable

# Більш читабельний спосіб (з дужками)
result = (first_variable + second_variable +
          third_variable + fourth_variable +
          fifth_variable)

Ключова відмінність полягає в тому, що всередині дужок Python автоматично розуміє, що вираз продовжується на наступному рядку, тому не потрібно додавати спеціальні символи. Це робить код чистішим та читабельним. Максимальна рекомендована довжина рядка - 80 символів.

## 2. Змінні
**Замість епіграфу**: *Коли дружина програміста народила двійню щоб не заморочуватись він назвав дітей `new_daughter1` і `new_daughter2`*

    Змінна (variable) — об'єкт програми, що має ім'я та значення.

Усі дані у комп'ютері зберігаються у пам'яті. Щоб отримати доступ до даних нам необхідно знати де саме у пам'яті вони знаходяться. Пам'ять комп'ютера поділена на комірки, і кожна комірка має свій номер. Отже у якості місця у пам'яті можна вказати її номер (адрес). На самому початку розвитку комп'ютерної техніки люди писали програми у машиних кодах, тобто у прямих інструкціях для процесора, і місцезнаходження даних так і визначали — вказуючи їх адрес у пам'яті. Це була дуже кропітка і складна робота, припуститись помилки було дуже легко. Згодом була написана програма Асемблер (Assembler, складальник), яка отримувала програму у текстовому вигляді і перекодувала її у машині коди для процесора. Програма складалась з інструкцій для процесора, але вже у більш зрозумілому для людини вигляді. А для звернення до певної комірки пам'яті замість адреси комірки почали використовувати мнемонічні імена. Це набагато полегшило процес розробки програм, а правила представлення машиних команд у текстовому вигляді назвали "мова Асемблера". Можна сказати що мова Асемблера — це перша мова програмування.

З розвитком мов програмування вони дедалі ставали усе вищого рівня, тобто такими, які ближче до людської мови ніж до команд для процесора. І для звернення до даних у пам'яті з'явилось таке поняття як змінна.

У різних мовах програмування принципи роботи зі змінними можуть дещо відрізнятись. Для початку припустимось ствердження що змінна — це певне ім'я, яке ми можемо пов'язати з певними даними, і для доступу до останніх ми можемо використовувати це ім'я.

### Змінні у Python
Щоб пов'язати дані з іменем використовують операцію присвоєння. В Python ця операція позначається знаком "дорівнює". Зліва від знаку "=" вказують ім'я (змінну), а праворуч — дані які треба пов'язати з цим ім'ям. Наприклад:

```python
quantity = 7
```
Цей вираз означає що до даних які дорівнюють числу 7 ми будемо звертатись за ім'ям `quantity`.
А читається цей вираз так: "Змінній `quantity` присвоєно значення `7`".

### Іменування змінних
Імена змінним у Python можна давати підпорядковуючись певним правилам:

- ім'я може містити букви, цифри, символ підкреслення (_)
- ім'я не може починатись з цифри
- ім'я не може співпадати з одним з ключових слів (словами які використовуються у мові програмування Python)
- імена чутливі до регістру
Спи#сок ключових слів можна дізнатись виконавши спеціальну команду:

In [8]:
help('keywords')


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 



Часто імена змінних включають у себе декілька слів. Наприклад, «ім'я користувача».

У різних мовах застосовують різні стилі іменування, ось основні:

`camelCase` — перше слово з маленької букви, усі наступні — з великої
`snake_case` — усі слова з маленької букви, розділяють символом підкреслення
`kebab-case` — усі слова з маленької букви, розділяють символом дефіса
`PascalCase` — усі слова з великої букви

Змінні в Python іменують в стилі `snake_case`. Усі слова — англійською мовою.

```python
user_name
```

Серед розробників є жарт: `«саме складне у програмуванні — назви змінних і інвалідація кеша»`. Придумати назву дійсно деколи складно. Як би ви назвали змінну, у якій зберігається кількість неоплачених замовлень від клієнтів, які мають заборгованість у минулому кварталі?

Але все ж намагайтесь давати змінним осмислені імена які відносяться до даних, на які вони вказують. З часом, вивчаючи чужий код, ви сформуєте правильні поняття іменування.

### Використання змінних
Ми можемо давати команди інтерпретатору присвоїти змінним певні значення. А щоб дізнатись значення змінної достатньо ввести її ім'я (в інтерактивній консолі або Jupiter ноутбуці) або розрукувати її через команду `print` (в скріпті)

Приклад:

In [9]:
usd_rate = 41.25
usd = 150
uah = usd_rate * usd

In [10]:
uah

6187.5

In [11]:
print(usd_rate)

41.25


У Python можна робити множинне присвоювання:

In [12]:
a = b = c = 8

In [13]:
print('a =', a)

a = 8


In [14]:
# # String Interpolation / f-Strings
print(f'{b = }')

b = 8


In [15]:
# “Old Style” String Formatting (% Operator)
print('c = %s' %(c))

c = 8


In [16]:
# “New Style” String Formatting (str.format)
print('a = {}; b = {}; c = {};'.format(a,b,c))

a = 8; b = 8; c = 8;


Коли ми присвоюємо змінній певне значення вперше, говорять що ми `проініціалізували` змінну, а така дія називається `ініціалізація`.

Щоб використовувати будь-яку змінну її треба спочатку ініціалізувати. Якщо ж ми спробуємо отримати значення непроініціалізованої змінної, то інтерпретатор повідомить нам що ми припустились помилки, або ж у термінах Python що "виникла вийняткова ситуація".

In [17]:
variable = 3
ans = variable * 10

Інформація про вияйняткову ситуацію починається рядком
```python
Traceback (most recent call last):
```

а як правило останній рядок пояснює нам що саме спричинило вийняткову ситуацію.

```python
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[18], line 2
      1 # variable = 3
----> 2 ans = variable * 10

NameError: name 'variable' is not defined
```

У даному випадку рядок
```python
NameError: name 'variable' is not defined
```
можна прочитати як:

    Помилка імені: ім'я 'variable' не визначене.

Тобто ми намагались отримати значення змінної a але не проініцілізували її перед цим.

Детальніше про вийняткові ситуації та що з ними робити коли вони виникають ми дізнаємось у подальшому.

### Константи
Деякі дані, такі як математичні постійні, ніколи не змінюються. Наприклад число "`пі`". Наближено воно завжди дорівнє `3.14` і не може змінюватись. Такі дані називають константами.

В Python немає спеціального механізму для визначення констант. Створення константи не відрізняється від створення змінної. Але є домовленість: константи прийнято іменувати буквами у верхньому регістрі.
```python
PI = 3.14159
```

## 3. Поняття об'єкту в Python

У мові Python **все є об'єктом** - числа, рядки, списки, навіть самі типи даних. Об'єкт - це область пам'яті, яка містить деяке значення та асоціаційовані з ним набори операцій.

`Об’єкт` – це абстракція, яка використовується для представлення даних у мові Python. Об’єкт є фундаментальним поняттям у мові програмування Python (і не тільки в Python). Об’єкт є синонімом до слова екземпляр, який містить деяке значення даних.

Будь-які операції (додавання, віднімання тощо) у мові Python виконуються над об’єктами.

Дані у вигляді об’єктів можуть бути представлені двома способами:

- вбудованими об’єктами, які предоставляються мовою Python;
- власноруч створеними об’єктами, з застосуванням конструкцій мови Python чи бібліотек розширень.

Використання вбудованих об’єктів дає наступні переваги:

- спрощується розробка програм. Не витрачаються зайві ресурси на розробку власних складних об’єктів. Достатньо використовувати наявні потужні вбудовані засоби мови Python;
- знижується кількість помилок на розробку власних об’єктів, оскільки програмний код вбудованих об’єктів є протестований;
- забезпечується достатньо висока швидкодія з обробки колекцій, списків, таблиць тощо. Це пояснюється тим, що програмний код вбудованих об’єктів є максимально оптимізований;
- можливість розробки власних складних об’єктів на базі вбудованих класів мови Python чи інтерфейсів мови C. Це в свою чергу полегшує розширюваність вже існуючого коду;
- вбудовані об’єкти реалізовані так, що вони є стандартною і незмінною частиною мови Python. Це дає перевагу зручності та високій ефективності їх використання у програмах. Власно розроблені об’єкти мають схильність змінюватись від одного випадку до іншого.

### 2.1 Характеристики об'єкту

Кожен об'єкт має три характеристики:

1. **Ідентичність** - унікальне число для кожного об'єкту
2. **Тип** - визначає можливі операції та значення
3. **Значення** - дані, які зберігає об'єкт

In [18]:
# Визначення характеристик об'єкту
x = 42

In [19]:
# Значення
print(x)            # 42

42


In [20]:
# Тип об'єкту
print(type(x))      # <class 'int'>

<class 'int'>


`Ідентичність об’єкту` – це ціле унікальне і постійне (константне) число, яке встановлюється для даного конкретного об’єкту. Ідентичність об’єкту встановлюється тільки один раз при створенні об’єкту. Ідентичність даного об’єкту ніколи не змінюється після його створення. Ідентичність об’єкту асоціюється з адресою об’єкту в пам’яті.

Функція `id()` призначена для отримання значення ідентичності об’єкту. Функція повертає ціле число, яке для даного об’єкту є унікальним та постійним.

In [21]:
# Ідентичність - унікальний ідентифікатор в пам'яті
print(id(x))        # наприклад: 140706012416200

140706026899656


In [22]:
print(id(42))

140706026899656


In [23]:
a=5
print(id(a))

140706026898472


In [24]:
pa=id(a)
print(pa)

140706026898472


In [25]:
print(id(5))

140706026898472


In [26]:
print(id(6))

140706026898504


У вищенаведеному прикладі визначається значення ідентичності для об’єктів з іменами `a`, `pa`, числа `6`, числа `5`. Як видно з прикладу, числа також є об’єктами. Значення ідентичності для різних імен можуть співпадати.

З допомогою операторів `is` та `is not` можна порівнювати значення об’єктів на ідентичність.

In [27]:
a=3
b=5
print(a is b)

False


In [28]:
c = 5
print(c is b)

True


In [29]:
print(c is not a)

True


In [30]:
x = 256
y = 256
print(x is y)

True


In [31]:
x = 257
y = 257
print(x is y)

False


In [32]:
print(x == y)

True


Це приклад того, як у Python працює оператор is та внутрішня оптимізація для цілих чисел.

🔍 Що відбувається
- is перевіряє чи це один і той самий об’єкт у пам’яті, а не чи значення однакові.
- Для малих цілих чисел (зазвичай від -5 до 256) Python створює їх наперед і зберігає в спеціальному кеші. Це називається інтернування (interning).
- Тому, коли ми пишемо:
```python
x = 256
y = 256
print(x is y)  # True
```
- обидві змінні посилаються на той самий об’єкт у пам’яті.
- Але для чисел поза цим діапазоном (наприклад, 257) Python не гарантує кешування:
```python
x = 257
y = 257
print(x is y)  # False
```
- У цьому випадку створюються два різні об’єкти з однаковим значенням, тому is повертає False.
  
💡 Висновок:
- Якщо потрібно перевіряти значення, використовуємо `==`.
- Якщо потрібно перевірити, чи це той самий об’єкт, використовуємо `is`.


In [33]:
int_5 = 5
float_5 = 5.0
print(int_5 == float_5)

True


In [34]:
print(int_5 is float_5)

False


Цей приклад показує різницю між **`==`** (перевірка рівності значень) та **`is`** (перевірка ідентичності об’єктів у пам’яті) у Python. 

---

#### 🔍 Що відбувається крок за кроком

```python
int_5 = 5
float_5 = 5.0
print(int_5 == float_5)  # True
```
- **`==`** перевіряє, чи значення однакові.  
- `5` (тип `int`) і `5.0` (тип `float`) мають однакове числове значення, тому результат **`True`**.  
- Python автоматично виконує порівняння з приведенням типів, якщо це можливо.

```python
print(int_5 is float_5)  # False
```
- **`is`** перевіряє, чи це **той самий об’єкт у пам’яті**.  
- `int_5` — це ціле число (`int`), а `float_5` — число з плаваючою комою (`float`).  
- Це **різні типи** і різні об’єкти, навіть якщо значення збігаються.  
- Тому результат **`False`**.

---

#### 📌 Підсумок
- **`==`** → порівнює **значення** (може бути True навіть для різних типів, якщо значення збігаються).  
- **`is`** → перевіряє **ідентичність об’єкта** (True тільки якщо змінні вказують на один і той самий об’єкт у пам’яті).  

---


## 4. Дані та їх типи
У реальному житті ми виконуємо різноманітні дії над оточуючими нас предметами (об'єктами). Ми змінюємо їх властивості, наділяємо новими функціями. По аналогії з цим комп'ютерні програми також маніпулюють об'єктами, тільки віртуальними, цифровими. Поки що будемо називати такі об'екти даними.

Вочевидь, дані бувають різними: числа, текстова інформація, аудіозапис. Числа у свою чергу також бувають різними: цілими, дійсними, можуть мати дуже велике значення або дуже довгу дробову частину.

    Тип даних — характеристика, яка визначає множину припустимих значень, формат їхнього збереження та набір операцій (деколи також розмір виділеної пам'яті), які можна робити над даними.

### Основні типи даних
Python має багато вбудованих типів даних. Ось деякі найважливіші:

* Числа можуть бути цілими (1 і 2), з десятковими дробами (1.1 і 1.2), звичайними дробами (½ and ⅔), чи навіть комплексними.
* Логічні (Булеві) змінні приймають значеня True або False.
* Рядки (послідовності символів, наприклад 'To be or not to be' або ж цілий HTML документ)
* Байти та масиви байтів (наприклад зображення в форматі JPEG)
* Списки (впорядковані послідовності значень).
* Кортежі (впорядковані незмінні послідовності значень).
* Множини (набори унікальних значень).
* Словники (набори пар ключ-значення).

<img src="data_types.svg" alt="Типи даних" width="400"/>

  
В Python тип даних не треба вказувати явно. Покладаючись на початково присвоєне змінній значення, Python визначає якому типу воно належить і запам’ятовує його самостійно. Таку типізацію називають динамічною або неявною.

### Динамічна типізація
Python є мовою з **динамічною типізацією** - тип змінної визначається автоматично при присвоєнні значення:

In [35]:
# Не потрібно оголошувати тип
x = 10          # x автоматично стає int
print(type(x))

<class 'int'>


In [36]:
x = "Python"    # тепер x стає str
print(type(x))

<class 'str'>


In [37]:
x = 3.14        # тепер x стає float
print(type(x))

<class 'float'>


Не зважаючи на це, буває корисно явно вказувати тип змінної використовуючи `анотацію типів`.
 
---

#### 📌 Що таке анотація типів
Анотація типів (type hints) — це спосіб **підказати** іншим програмістам (і інструментам, як-от IDE чи лінтери), які типи даних очікуються у змінних, аргументах функцій та значеннях, що повертаються.  
Вона **не змінює поведінку програми** під час виконання, але допомагає:
- робити код зрозумілішим;
- знаходити помилки ще до запуску;
- полегшувати автодоповнення в редакторах.

---

#### 🔹 Приклади з простими типами

```python
age: int = 25        # ціле число
price: float = 19.99 # число з плаваючою комою
product: str = "Apple" # рядок
```
Тут ми вказали, що:
- `age` має бути цілим числом (`int`);
- `price` — десятковим (`float`);
- `product` — рядком (`str`).

#### 💡 Корисно знати
- Python **не примушує** дотримуватися анотацій — це лише підказка.
- Для перевірки типів можна використовувати інструменти, як-от **mypy**:
  ```bash
  mypy script.py
  ```
---

#### Базові типи
| Тип | Приклад значення | Анотація |
|-----|------------------|----------|
| `int` | `42` | `age: int` |
| `float` | `3.14` | `pi: float` |
| `str` | `"Hello"` | `name: str` |
| `bool` | `True` | `is_active: bool` |
| `None` | `None` | `value: None` |

---

### Операції у програмуванні
`Операція` – це виконання певних дій над даними. Дані, над якими виконують дій, називають `операндами`. Саму дію виконує `оператор` – спеціальний інструмент.

Наприклад додавання чисел:
```python
2 + 3
```
Тут операцією виступає додавання, операндами — числа `2` і `3`, оператором — знак "`+`".

Зауважимо, що `один і той же оператор може виконувати різні дії (операції) залежно від типу даних операндів`.

In [38]:
number1 = 20
number2 = 30
res_number = number1 + number2
print(res_number)

50


In [39]:
string1 = '20'
string2 = '30'
res_string = string1 + string2
print(res_string)

2030


### Числа
Python підтримує як цілі, так і дробові числа. Вказувати тип числа явно не потрібно, Python сам визначить його за наявністю чи відсутністю десяткової крапки.

### Цілі числа (int)
Числа в Python нічим не відрізняються від звичайних чисел. 

In [40]:
# Цілі числа (int)
age = 25
temperature = -10
big_number = 1_000_000  # можна використовувати _ для читабельності

In [41]:
print(big_number)

1000000


In [42]:
print(f"{big_number:_}")

1_000_000


Цілі числа в Python підтримують набір самих звичайних математичних операцій:
| операція | пояснення                     |
|----------|-------------------------------|
| x + y    | додавання                     |
| x - y    | віднімання                    |
| x * y    | множення                      |
| x / y    | ділення                       |
| x // y   | ціла частина від ділення      |
| x % y    | залишок від ділення           |
| -x       | зміна знаку числа             |
| x ** y   | підведення у ступінь          |

Операції мають пріоритет який можна змінювати використовуючи круглі дужки, усе так само як у математиці.

In [43]:
a = 4 + 3 * 2
b = (4 + 3) * 2
print(f'{a = }')
print(f'{b = }')

a = 10
b = 14


Зауважте, що результатом операції `/`, а також підведення у від'ємний або дійсний ступінь буде дійсне число.

In [44]:
print(20 / 3)

6.666666666666667


In [45]:
print(20 // 3)

6


In [46]:
print(20 % 3)

2


In [47]:
print(3 ** 4)

81


In [48]:
print(81 ** (1/4))

3.0


Також слід зазначити, що цілі числа в Python, на відміну від багатьох інших мов програмування, підтримують довгу арифметику.
В Python довга (або довільна) арифметика реалізована вбудовано для цілих чисел (тип `int`). Вам не потрібно писати спеціальний код для обробки дуже великих чисел, оскільки тип `int` автоматично підтримує числа будь-якого розміру. Просто використовуйте стандартні оператори, а Python автоматично керує розміром цілих чисел, дозволяючи їм зростати необмежено, на відміну від мов з фіксованою розмірністю цілих чисел. 

In [49]:
print(3 ** 150)

369988485035126972924700782451696644186473100389722973815184405301748249


Вище ми бачимо приклад того, як Python працює з **дуже великими цілими числами** — те, що називається **"arbitrary-precision integers"**.

---

### 🔍 Що відбувається в коді:

```python
print(8 ** 119)
```

- Підносимо число `8` до степеня `119`.
- Це створює **гігантське число**:  
  `369806799885152159729427087251456644481673063392722957731054430429731194129`

---

### 📐 Чому це можливо в Python?

На відміну від багатьох мов програмування (наприклад, C або Java), де цілі числа мають обмежений розмір (наприклад, `int32` або `int64`), Python автоматично переходить на **довільну точність**:

- Немає переповнення (`overflow`) — Python просто виділяє більше пам’яті.
- Тип `int` в Python — це **динамічна структура**, яка масштабується під потреби.

---

### 📊 Порівняння з іншими мовами:

| Мова       | Максимальний `int` (типовий) | Поведінка при переповненні |
|------------|------------------------------|-----------------------------|
| C          | 2,147,483,647 (`int32`)      | Переповнення, неправильне значення |
| Java       | 9,223,372,036,854,775,807 (`long`) | Переповнення |
| Python     | Без обмежень (`int`)         | Автоматичне розширення     |

---

### 🧠 Навіщо це потрібно?

- Криптографія (робота з великими простими числами)
- Наукові обчислення
- Теорія чисел
- Фінансові розрахунки з високою точністю

---

### Дійсні числа (float)
У дійсних числах ціла частина від дробової відділяється знаком крапки.

In [50]:
# Дійсні числа
height = 175.5
temperature = -15.7
pi = 3.14159265359

Для запису дуже великих або дуже малих по модулю чисел використовують так званий запис «з плаваючою крапкою» (також науковий або інженерний запис). У цьому випадку число представляється у вигляді деякої десяткової дроби, яку називають мантисою, помноженої на цілочислений ступінь десяти (порядок). Наприклад, відстань від Землі до Сонця дорівнює `1.496×10¹¹`.

Числа с плавачою крапкою записуються так: спочатку пишеться `мантиса`, далі пишеться буква `e`, потім пишеться `порядок`. Пробіли всередині цього запису не ставляться.
```python
1.496e11
```

In [51]:
# Наукова нотація
scientific = 1.5e6   # 1500000.0
small = 2.5e-3       # 0.0025

In [52]:
print(f'{scientific = }')
print(f'{small = }')

scientific = 1500000.0
small = 0.0025


Дробові числа підтримують ті ж операції, що й цілі. Однак через особливості представлення чисел у пам'яті комп'ютера дробові числа неточні, це може спричиняти помилки при обчисленнях:

In [53]:
print(0.1 + 0.2)

0.30000000000000004


In [54]:
print(1/10 + 1/5)

0.30000000000000004


Це класична проблема **неточності з плаваючою комою** (floating-point precision), яка виникає через те, як комп’ютери представляють десяткові числа у **бінарному форматі**.

---

### 🔍 Що саме відбувається?

Коли ми виконуємо:

```python
print(0.1 + 0.2)
```

Очікуємо побачити `0.3`, але отримуємо `0.30000000000000004`. Причина — **деякі десяткові дроби не можуть бути точно представлені у двійковій системі**, так само як 1/3 не може бути точно записана у десятковій (0.333...).

У двійковій системі `0.1` і `0.2` — це нескінченні періодичні дроби, які комп’ютер **обрізає** до певної точності. Коли ці наближені значення додаються — виникає маленька похибка.

---

### 🧠 Чому це важливо?

- У фінансових розрахунках, фізичних симуляціях або тестах — навіть мікроскопічна похибка може призвести до неправильних результатів або провалу тестів.
- Це не баг Python — це **обмеження стандарту IEEE 754**, який використовують майже всі мови програмування.

---

Для високої точності використовують інші засоби (наприклад `Decimal` і `Fraction`)
Ось практичні приклади використання `Decimal` і `Fraction` — двох потужних інструментів з модуля `decimal` та `fractions`, які дозволяють уникнути проблем з точністю при роботі з числами.

### 🔢 `Decimal`: для точних десяткових обчислень

In [55]:
from decimal import Decimal, getcontext

getcontext().prec = 10  # встановлюємо точність до 10 знаків

a = Decimal('0.1')
b = Decimal('0.2')
c = a + b

print(c)  # 0.3

0.3


📌 **Переваги:**
- Точне представлення десяткових дробів
- Контроль над кількістю знаків після коми
- Ідеально для фінансових розрахунків

### 🧮 `Fraction`: для точних раціональних чисел

In [56]:
from fractions import Fraction

a = Fraction(1, 10)  # 0.1
print(a)

1/10


In [57]:
b = Fraction(1, 5)   # 0.2
print(b)

1/5


In [58]:
c = a + b
print(c)  # 3/10

3/10


In [59]:
print(type(c))

<class 'fractions.Fraction'>


In [60]:
print(float(c))  # 0.3

0.3


In [61]:
print(c + 1)

13/10


In [62]:
print(c + 0.1)

0.4


📌 **Переваги:**
- Зберігає точне співвідношення чисел (наприклад, 1/3)
- Уникає округлення до float
- Добре підходить для математичних задач, де важлива точність дробів

### 🔍 Порівняння:

| Засіб      | Тип точності       | Приклад використання         | Підходить для              |
|------------|--------------------|-------------------------------|----------------------------|
| `float`    | обмежена, неточна  | `0.1 + 0.2` → `0.30000000004` | загальні обчислення        |
| `Decimal`  | точна десяткова    | `Decimal('0.1') + Decimal('0.2')` → `0.3` | фінанси, звітність         |
| `Fraction` | точна раціональна | `Fraction(1, 10) + Fraction(1, 5)` → `3/10` | математика, теорія чисел   |

Також дійсні числа не підтримують довгу арифметику.

In [63]:
print(1/2 ** 50)

8.881784197001252e-16


In [64]:
print(2.0 + 2.0)

4.0


In [65]:
print(2 + 2.0)

4.0


#### `complex` (комплексні числа)
Для математичних обчислень з комплексними числами:

In [66]:
# Комплексні числа
complex1 = 2 + 3j
print(complex1)

(2+3j)


In [67]:
print(type(complex1))

<class 'complex'>


In [68]:
complex2 = complex(4, -2)  # 4-2j
print(complex2)

(4-2j)


In [69]:
# Доступ до частин
print(complex2.real)   # 2.0 - дійсна частина

4.0


In [70]:
print(complex2.imag)   # 3.0 - уявна частина

-2.0


In [74]:
import cmath

complex3 = cmath.sqrt(-4)
print(complex3)

2j


### Логічний тип (bool)

Може приймати лише два значення:

```python
# Булеві значення
is_student = True
is_finished = False

# Перетворення в bool
print(bool(1))      # True - будь-яке число крім 0
print(bool(0))      # False
print(bool(""))     # False - порожній рядок
print(bool("Hi"))   # True - непорожній рядок

In [ ]:
# Булеві значення
is_student = True
is_finished = False

In [75]:
# Перетворення в bool
print(bool(1))      # True - будь-яке число крім 0

True


In [76]:
print(bool(0))      # False

False


In [77]:
print(bool(""))     # False - порожній рядок

False


In [78]:
print(bool("Hi"))   # True - непорожній рядок

True


In [80]:
print(2<3)

True


In [81]:
print(4>5)

False


In [79]:
print(bool(None))

False


### `None` в Python: сутність "нічого"

В світі програмування на Python `None` — це спеціальний та унікальний об'єкт, який слугує для позначення відсутності значення. Це спосіб сказати, що змінна, параметр або результат функції не мають конкретного значення. `None` є єдиним екземпляром свого власного типу даних, `NoneType`.

Уявіть собі коробку. Вона може містити щось — число, рядок, список. А може бути порожньою. Ось ця "порожнеча" і є `None` в Python. Змінна існує, але вона не вказує на жоден конкретний об'єкт.

### Основні характеристики `None`:

  * **Єдиний екземпляр:** Всі посилання на `None` вказують на один і той самий об'єкт в пам'яті. Це робить перевірку на `None` дуже ефективною.
  * **Тип `NoneType`:** `None` є єдиним можливим значенням типу `NoneType`.
  * **"Хибне" значення (False):** В логічному контексті `None` поводиться як `False`. Це означає, що воно не пройде перевірку в умовному операторі `if`.

### Як правильно перевіряти на `None`?

Найкращий спосіб перевірити, чи є змінна `None` — це використання оператора `is`.

In [82]:
variable = None
print(variable is None)

True


**Чому `is`, а не `==`?**

Хоча `змінна == None` в більшості випадків спрацює так само, як `is`, використання `is` є більш "пітонічним" (Pythonic) і надійним. Оператор `is` перевіряє, чи дві змінні вказують на один і той самий об'єкт в пам'яті. Оскільки `None` є унікальним об'єктом, ця перевірка є швидкою та гарантує, що ви порівнюєте саме з ним. Теоретично, об'єкт може бути створений таким чином, щоб `== None` повертало `True`, навіть якщо цей об'єкт не є справжнім `None`.

### `None`, `False` та `0`: в чому різниця?

Хоча всі три значення вважаються "хибними" (falsy) в логічному контексті (тобто `bool(None)`, `bool(False)` і `bool(0)` повернуть `False`), вони не є однаковими:

| Значення | Тип          | Опис                                  |
| :------- | :----------- | :------------------------------------ |
| `None`   | `NoneType`   | Представляє відсутність значення.     |
| `False`  | `bool`       | Представляє логічне значення "хиба".  |
| `0`      | `int` (або `float`) | Представляє числове значення "нуль". |

In [83]:
print(None == False)  # Виведе: False

False


In [84]:
print(None == 0)      # Виведе: False

False


In [85]:
print(False == 0)     # Виведе: True (з історичних причин)

True


Важливо розрізняти їх. Наприклад, функція може повернути `0`, що є валідним числовим результатом, або `None`, що може сигналізувати про помилку або відсутність даних. Використання `if` не дозволить розрізнити ці два випадки. Тому завжди краще явно перевіряти на `None` за допомогою `is`.

Автоматичне приведення типів для чисел у Python — це один з найяскравіших прикладів type coercion. Python автоматично перетворює числові типи для забезпечення сумісності операцій.

#### Ієрархія числових типів

Python дотримується чіткої ієрархії при приведенні числових типів:

```
bool → int → float → complex
```

**Правило:** менш "складний" тип завжди приводиться до більш "складного".

In [ ]:
# bool + int
result = True + 5
print(result)        # 6
print(type(result))  # <class 'int'>

In [ ]:
result = False * 10
print(result)        # 0
print(type(result))  # <class 'int'>

In [ ]:
# int + float
result = 3 + 2.5
print(result)        # 5.5
print(type(result))  # <class 'float'>

In [ ]:
result = 10 / 3      # Ділення завжди повертає float
print(result)        # 3.3333333333333335
print(type(result))  # <class 'float'>

In [ ]:
# float + complex
result = 2.5 + (1 + 2j)
print(result)        # (3.5+2j)
print(type(result))  # <class 'complex'>

In [ ]:
result = 3.14 * 2j
print(result)        # 6.28j
print(type(result))  # <class 'complex'>

#### Комбіновані операції

In [ ]:
# bool + int + float + complex
result = True + 2 + 3.5 + (1 + 1j)
print(result)        # (7.5+1j)
print(type(result))  # <class 'complex'>

#### Операції порівняння

При порівнянні Python також виконує приведення типів:

In [ ]:
print(1 == 1.0)      # True

In [ ]:
print(True == 1)     # True

In [ ]:
print(False == 0.0)  # True

In [ ]:
print(2 == (2+0j))   # True

In [ ]:
# Порівняння величини
print(1 < 2.5)       # True

In [ ]:
print(True < 3.14)   # True

#### Детальний розбір прикладів

##### Приклад 1: Послідовне приведення

In [ ]:
a = True     # bool
b = 5        # int  
c = 2.3      # float
d = 1 + 1j   # complex

step1 = a + b        # True + 5 → 1 + 5 = 6 (int)
step2 = step1 + c    # 6 + 2.3 → 6.0 + 2.3 = 8.3 (float)  
step3 = step2 + d    # 8.3 + (1+1j) → (8.3+0j) + (1+1j) = (9.3+1j) (complex)

print(f"Фінальний результат: {step3}")  # (9.3+1j)
print(f"Тип: {type(step3)}")             # <class 'complex'>

#### Приклад 2: Математичні функції

In [ ]:
import math

# Функції можуть приймати різні числові типи
print(math.sqrt(4))      # 2.0 (int → float)

In [ ]:
print(math.sqrt(4.0))    # 2.0 (вже float)

In [ ]:
print(math.sqrt(True))   # 1.0 (bool → float)

In [ ]:
# Степінь може повертати різні типи
print(2 ** 3)            # 8 (int)

In [ ]:
print(2 ** 3.0)          # 8.0 (float)

In [ ]:
print(2.0 ** 3)          # 8.0 (float)

#### Спеціальні випадки

##### Ділення

In [ ]:
# Звичайне ділення завжди повертає float
print(6 / 2)             # 3.0 (не 3!)
print(type(6 / 2))       # <class 'float'>

In [ ]:
# Цілочисельне ділення зберігає тип
print(6 // 2)            # 3
print(type(6 // 2))      # <class 'int'>

In [ ]:
# Але може приводити до float
print(7.0 // 2)          # 3.0
print(type(7.0 // 2))    # <class 'float'>

Автоматичне приведення числових типів у Python робить мову гнучкою та зручною, але розуміння цих механізмів допомагає писати більш передбачуваний код.